src/hera.cpp

```diff
@@ -107,19 +107,26 @@ EXPORT struct evm_result evm_execute(struct evm_instance* instance,
 
   try {
     hera.execute(call);
+  } catch (OutOfGasException) {
+    ret.gas_left = EVM_EXCEPTION;
+  } catch (InternalErrorException &e) {
+    ret.gas_left = EVM_EXCEPTION;
+    std::cerr << "InternalError: " << e.what() << std::endl;
   } catch (std::exception &e) {
-    // FIXME: `evm_result` should have a way to report this back
-    Fatal() << "Execution failed: " << e.what() << "\n";
+    ret.gas_left = EVM_EXCEPTION;
+    std::cerr << "Unknown exception: " << e.what() << std::endl;
   }
 
-  // copy call result
-  ret.output_size = call->returnValue.size();
-  ret.output_data = (const uint8_t *)malloc(ret.output_size);
-  // FIXME: properly handle memory allocation issues
-  if (ret.output_data) {
-    std::copy(call->returnValue.begin(), call->returnValue.end(), (char *)ret.output_data);
+  if (ret.gas_left != EVM_EXCEPTION) {
+    // copy call result
+    ret.output_size = call->returnValue.size();
+    ret.output_data = (const uint8_t *)malloc(ret.output_size);
```

Comment:
use calloc() instead for safety; uninitialized memory can cause some nasty problems.

Reply:
True, but the entire buffer is overridden below. This assumes `std::copy` is working as intended.